### PR TITLE
Storing synced tips in DB

### PR DESCRIPTION
**Pull request by jmozah**
The PR supports the storing and retrieving of the `synced_tips` from the DB.
It supports two functions, 
- `UpdateValidatedTips`: This should be called whenever there is a change in the `synced_tips` data structure. 
- `ValidatedTips`: This can be used to retrieve all the `synced_tips` from the DB. Should be used during the start of the beacon node, to load the memory version of the data structure with the last known tips.

If the changes are fine, I would add these in the respective places in ta later PR.

### CODE REVIEW

File: beacon-chain/db/kv/validated_tips.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	oldVals, err := s.ValidatedTips(ctx)
+	if err != nil {
+		return err
+	}
+
+	updateErr := s.db.Update(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(validatedTips)
+
+		// Delete keys that are present and not in the new set.
+		for k, _ := range oldVals {
```
Comment (potuz) [inline beacon-chain/db/kv/validated_tips.go:49]:
```suggestion
		for k := range oldVals {
```

File: beacon-chain/db/kv/validated_tips_test.go [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	db := setupDB(t)
+
+	oldTips := make(map[[32]byte]types.Slot)
+	oldTips[[32]byte{'A'}] = types.Slot(1)
+	oldTips[[32]byte{'B'}] = types.Slot(2)
+	oldTips[[32]byte{'C'}] = types.Slot(3)
+
+	require.NoError(t, db.UpdateValidatedTips(ctx, oldTips))
+
+	// create a new overlapping tips to add
```
Comment (potuz) [inline beacon-chain/db/kv/validated_tips_test.go:41]:
```suggestion
	// create new non overlapping tips to add
```

**Review by potuz — COMMENTED** [review]
Looks Good to me.

File: beacon-chain/db/kv/validated_tips.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.UpdateValidatedTips")
+	defer span.End()
+
+	// Get the already existing tips.
+	oldVals, err := s.ValidatedTips(ctx)
+	if err != nil {
+		return err
+	}
+
+	updateErr := s.db.Update(func(tx *bolt.Tx) error {
```
Comment (terencechain) [inline beacon-chain/db/kv/validated_tips.go:45]:
```suggestion
	err := s.db.Update(func(tx *bolt.Tx) error {
```

File: beacon-chain/db/kv/validated_tips.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+		bkt := tx.Bucket(validatedTips)
+
+		// Delete keys that are present and not in the new set.
+		for k, _ := range oldVals {
+			if _, ok := newVals[k]; !ok {
+				deleteErr := bkt.Delete(k[:])
```
Comment (terencechain) [inline beacon-chain/db/kv/validated_tips.go:51]: Perhaps I am missing something, I don't understand why deletion is necessary here. If the input data is in `key` `value` form, a simple `bkt.Put(key, value)` shall replace the previous `value`

Comment (potuz) [reply beacon-chain/db/kv/validated_tips.go:51]: this is deleting the keys that are not in the updated version

Comment (terencechain) [reply beacon-chain/db/kv/validated_tips.go:51]: So this is all or nothing. Why go so far to store {`[32]byte` -> `types.Slot`} instead we could just do just use a fixed item key. {`[]byte("sync-tips-key")` -> `map[[32]byte]types.Slot`}. The latter seems easier to reason and better performance

Comment (terencechain) [reply beacon-chain/db/kv/validated_tips.go:51]: After more discussions, we concluded that the map would have a single entry most of the time, and we will be changing that entry. Keeping it this way is simpler

**Review by terencechain — APPROVED** [review]

File: beacon-chain/db/kv/validated_tips_test.go [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+func areTipsSame(got map[[32]byte]types.Slot, required map[[32]byte]types.Slot) bool {
+	if len(got) != len(required) {
+		return false
+	}
+
+	for k, v := range got {
+		if val, ok := required[k]; ok {
+			if !bytes.Equal(bytesutil.SlotToBytesBigEndian(v), bytesutil.SlotToBytesBigEndian(val)) {
```
[the remark this answers is not in the document]
Comment (jmozah) [reply beacon-chain/db/kv/validated_tips_test.go]: Changed.

**Review by terencechain — APPROVED** [review]

**Pull request merged**